### PR TITLE
[mono] Hot Reload: support for reloadable types

### PR DESCRIPTION
**Pull request by lambdageek**
Fully implement support for the `NewTypeDefinition` EnC capability: a hot reload edit can add a new type definition (class, struct, interface, enum) that contains new fields, methods, properties, events, and nested classes, and can have generic params.  Also add reflection support for all of the above.

This is sufficient to support hot reload of types tagged with [`CreateNewOnMetadataUpdateAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.createnewonmetadataupdateattribute?view=net-6.0).

Contributes to https://github.com/dotnet/runtime/issues/63643 and https://github.com/dotnet/runtime/issues/57365

---

The implementation introduces `MonoAddedDefSkeleton` which is a bookkeeping structure that records the metadata indexes of where to find methods, fields, properties and events for a newly-added class.  The issue is that normally `mono_class_create_from_typedef` expects to find all that stuff in a contiguous block starting at the row pointed by the `MONO_TYPEDEF_FIELD_LIST` or `MONO_TYPEDEF_METHOD_LIST` columns of the new typedef.  But in EnC deltas, those columns are zeroed out, and instead the EnCLog functions like `ENC_FUNC_ADD_METHOD` and `ENC_FUNC_ADD_FIELD` explicitly record when a new row is added that is relevant to a particular type definition.

So during the pass over the EnCLog, we start creating skeletons when we see a new row added to the typedef table.  Then when we see the add functions, we record the row indices of the field, method, etc tables.  As far as I can tell, the rows for a particular type definition are contiguous, so we just record the first index and the count.

At the end of the pass, we save the skeletons on the metadata delta, and when code finally creates the `MonoClass` for a particular freshly-added type definition, we check the skeleton and use the recorded rows to populate the fields and methods.  Event and property data is created on-demand (the demand being reflection) in basically the same way.

One important note: we try very hard not to accidentally materialize a newly-added class as a `MonoClass` during the update itself - we need to at least get through the entire EnCLog - otherwise we might not see every field/method/etc and set up the class incorrectly.

The upshot is that the newly-added `MonoClass` behaves like any other "normal" class - all its fields are laid out normally (they're stored in the object, not in a separate hash table).  It can be generic.  It can have base types, interfaces, etc.

This is different from how added fields, props and events will be implemented on existing classes - we won't modify `MonoClass:fields` or the `MonoClassPropertyInfo:properties` arrays - once they have been allocated, we don't want to mess with them.  Instead additions to existing classes with create `MonoClassMetadataUpdateField` (`MonoClassMetadataUpdateProperty` and `MonoClassMetadataUpdateEvent`) structs that are going to be stored on the `MonoClassMetadataUpdateInfo` that's associated with each `MonoClass`.  The various iterators like `mono_class_get_fields` and `mono_class_get_props` will be modified to return the added fields/props/etc after the existing ones.

This is already implemented for fields.  Props and Events will be part of a separate effort to implement reflection support for them.

### CODE REVIEW

File: src/mono/mono/metadata/class-internals.h [hunk 1]
```diff
@@ -193,6 +193,9 @@ struct _MonoProperty {
 	MonoMethod *get;
 	MonoMethod *set;
 	guint32 attrs;
+	/* added by metadata-update after class was created;
+	 * not in MonoClassPropertyInfo array - don't do ptr arithmetic */
+	guint32 from_update : 1;
```
Comment (vargaz) [inline src/mono/mono/metadata/class-internals.h:198]: Can't these be stored in attrs ?

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/metadata/class-internals.h (hunk 1 above)
Comment (lambdageek) [reply src/mono/mono/metadata/class-internals.h:198]: @vargaz it's the same as fields - I really don't think there's an unused bit in there.

Comment (lambdageek) [reply src/mono/mono/metadata/class-internals.h:198]: Hm. Actually I'm wrong. Looks like only the low 16 bits have any sort of value (with a big chunk for properties just marked "unused").  So maybe I can put it in the high 16 bits

Comment (lambdageek) [reply src/mono/mono/metadata/class-internals.h]: Ok pushed https://github.com/dotnet/runtime/pull/66749/commits/9917ef3c704104da0c6f2e95710e6f7853f7331d which reserves a bit for hot reload.

**Pull request merged**